core/src/test/java/apoc/trigger/TriggerClusterRoutingTest.java

```diff
@@ -0,0 +1,71 @@
+import apoc.util.TestcontainersCausalCluster;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.neo4j.driver.types.Node;
+import org.neo4j.internal.helpers.collection.MapUtil;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static apoc.trigger.TriggerDeprecatedProcedures.SYS_NON_LEADER_ERROR;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class TriggerClusterRoutingTest {
+
+    private static TestcontainersCausalCluster cluster;
+
+    @BeforeClass
+    public static void setupCluster() {
+        cluster = TestContainerUtil
+                .createEnterpriseCluster(3, 1, Collections.emptyMap(), MapUtil.stringMap(
+                        "NEO4J_dbms_routing_enabled", "true",
+                        "apoc.trigger.enabled", "true"
+                ));
+    }
+
+    @AfterClass
+    public static void bringDownCluster() {
+        if (cluster != null) {
+            cluster.close();
+        }
+    }
+
+    @Test
+    public void testTriggerAddAllowedOnlyInLeaderMember() {
+        final List<Neo4jContainerExtension> members = cluster.getClusterMembers();
+        assertEquals(4, members.size());
+        for (Neo4jContainerExtension container: members){
+            // we skip READ_REPLICA members
+            final String readReplica = TestcontainersCausalCluster.ClusterInstanceType.READ_REPLICA.toString();
+            if (readReplica.equals(container.getEnvMap().get("NEO4J_dbms_mode")) || container.getSession() == null) {
+                continue;
+            }
+            final String systemRole = TestContainerUtil.singleResultFirstColumn(container.getSession(), "CALL dbms.cluster.role('system')");
+            if ("LEADER".equals(systemRole)) {
+                container.getSession().run("CALL apoc.trigger.add($name, 'UNWIND $createdNodes AS n SET n.ts = timestamp()',{})",
+                        Map.of("name", "trigger-" + container.getContainerName()));
+
+                container.getSession().run("CREATE (f:Foo)");
+                TestContainerUtil.testCall(container.getSession(), "MATCH (f:Foo) RETURN f",
+                        (row) -> assertTrue(((Node) row.get("f")).containsKey("ts")));
+            } else {
+                try {
+                    TestContainerUtil.testCall(container.getSession(), "CALL apoc.trigger.add($name, 'UNWIND $createdNodes AS n SET n.ts = timestamp()',{})",
+                            Map.of("name", "trigger-" + container.getContainerName()),
+                            row -> fail("Should fail because of non leader trigger addition"));
+                } catch (RuntimeException e) {
+                    String errorMsg = e.getMessage();
+                    assertTrue("The actual message is: " + errorMsg, errorMsg.contains(SYS_NON_LEADER_ERROR));
+                }
+            }
+        }
+    }
```

Comment:
I would really have liked this test to be written in terms of how a user should structure his queries using the Java driver in production.

Since users don't have access to container information then we shouldn't be using that.

I don't have the exact syntax but I would have liked to basically see us open a new driver session against "system" and create a write transaction which should get automatically routed to the "system" database LEADER because it's a write. This routing should happen automatically in the client if client-side routing enabled, if not it should happen automatically in the backend (in the DBMS) if server-side routing is enabled.

Then, creating another session the "neo4j" database, I would expect the `CREATE (f: Foo)` statement to be wrong.

Reply:
At the very least, this current implementation of the test has two problems:
- adding trigger is asynchronous but the test ignores that and will be very flaky as a result
- the user query which creates FOO is run against the leader of the system database which will often not be the leader of the user database, which will also be a source of flakiness